File: src/main/java/org/gecko/view/contextmenu/SystemViewElementContextMenuBuilder.java
package org.gecko.view.contextmenu;

import javafx.scene.control.ContextMenu;
import javafx.scene.control.MenuItem;
import javafx.scene.control.SeparatorMenuItem;
import org.gecko.actions.ActionManager;
import org.gecko.view.views.shortcuts.Shortcuts;
import org.gecko.viewmodel.SystemViewModel;

/**
 * Represents a type of {@link ViewContextMenuBuilder} for a {@link ContextMenu} specific to a
 * {@link org.gecko.view.views.viewelement.SystemViewElement SystemViewElement}. Contains {@link MenuItem}s that run
 * operations like opening or deleting the system.
 */
public class SystemViewElementContextMenuBuilder extends ViewContextMenuBuilder {

    private final SystemViewModel systemViewModel;

    private static final String OPEN_SYSTEM_MENU_ITEM = "Open System";

    public SystemViewElementContextMenuBuilder(
        ActionManager actionManager, SystemViewModel systemViewModel) {
        super(actionManager);

        this.systemViewModel = systemViewModel;
    }

    @Override
    public ContextMenu build() {
        ContextMenu systemContextMenu = super.build();

        SeparatorMenuItem dataTransferToSystemAccessSeparator = new SeparatorMenuItem();

        // Access system commands:
        MenuItem openSystemMenuItem = new MenuItem(OPEN_SYSTEM_MENU_ITEM);
        openSystemMenuItem.setOnAction(
            e -> actionManager.run(actionManager.getActionFactory().createViewSwitchAction(systemViewModel, false)));
        openSystemMenuItem.setAccelerator(Shortcuts.OPEN_CHILD_SYSTEM_EDITOR.get());

        MenuItem deleteMenuItem = new MenuItem(DELETE_MENU_ITEM);
        deleteMenuItem.setOnAction(e -> actionManager.run(
            actionManager.getActionFactory().createDeletePositionableViewModelElementAction(systemViewModel)));


        systemContextMenu.getItems().addAll(dataTransferToSystemAccessSeparator, openSystemMenuItem, deleteMenuItem);
        return systemContextMenu;
    }
}

File: src/main/kotlin/org/gecko/view/contextmenu/SystemViewElementContextMenuBuilder.kt
package org.gecko.view.contextmenu

import javafx.event.ActionEvent
import javafx.event.EventHandler
import javafx.scene.control.ContextMenu
import javafx.scene.control.MenuItem
import javafx.scene.control.SeparatorMenuItem
import org.gecko.actions.ActionManager
import org.gecko.view.GeckoView
import org.gecko.view.ResourceHandler
import org.gecko.view.views.shortcuts.Shortcuts
import org.gecko.viewmodel.SystemViewModel

/**
 * Represents a type of [ViewContextMenuBuilder] for a [ContextMenu] specific to a
 * [SystemViewElement][org.gecko.view.views.viewelement.SystemViewElement]. Contains [MenuItem]s that run
 * operations like opening or deleting the system.
 */
class SystemViewElementContextMenuBuilder(
    actionManager: ActionManager,
    val systemViewModel: SystemViewModel,
    geckoView: GeckoView
) : ViewContextMenuBuilder(actionManager, geckoView) {
    override fun build(): ContextMenu {
        val systemContextMenu = super.build()

        val dataTransferToSystemAccessSeparator = SeparatorMenuItem()

        // Access system commands:
        val openSystemMenuItem = MenuItem(ResourceHandler.Companion.open_system)
        openSystemMenuItem.onAction = EventHandler { e: ActionEvent? ->
            actionManager.run(
                actionManager.actionFactory.createViewSwitchAction(
                    systemViewModel, false
                )
            )
        }
        openSystemMenuItem.accelerator = Shortcuts.OPEN_CHILD_SYSTEM_EDITOR.get()

        val deleteMenuItem = MenuItem(ResourceHandler.Companion.delete)
        deleteMenuItem.onAction = EventHandler { e: ActionEvent? ->
            actionManager.run(
                actionManager.actionFactory.createDeletePositionableViewModelElementAction(systemViewModel)
            )
        }


        systemContextMenu!!.items.addAll(dataTransferToSystemAccessSeparator, openSystemMenuItem, deleteMenuItem)
        return systemContextMenu
    }
}
